src/connectors/onchain_events/mod.rs

```diff
@@ -0,0 +1,361 @@
+
+    #[error(transparent)]
+    UnableToParseUrl(#[from] url::ParseError),
+
+    #[error(transparent)]
+    UnableToParseLog(#[from] alloy_sol_types::Error),
+
+    #[error("Log missing block hash")]
+    LogMissingBlockHash,
+
+    #[error("Log missing log index")]
+    LogMissingLogIndex,
+
+    #[error("Log missing block number")]
+    LogMissingBlockNumber,
+
+    #[error("Log missing tx index")]
+    LogMissingTxIndex,
+
+    #[error("Unable to find block by hash")]
+    UnableToFindBlockByHash,
+}
+
+#[derive(Debug)]
+pub enum SignerEvent {
+    Add {
+        key: Bytes,
+        key_type: u32,
+        metadata: Bytes,
+        metadata_type: u8,
+    },
+    Remove {
+        key: Bytes,
+    },
+    AdminReset {
+        key: Bytes,
+    },
+}
+
+#[derive(Debug)]
+pub struct SignerMigratedEvent {
+    migrated_at: u64,
+}
+
+#[derive(Debug)]
+pub enum IdRegisterEvent {
+    Register {
+        to: Address,
+        recovery_address: Address,
+    },
+    Transfer {
+        from: Address,
+        to: Address,
+    },
+    ChangeRecovery {
+        recovery_address: Address,
+    },
+}
+
+#[derive(Debug)]
+pub struct StorageRentEvent {
+    payer: Address,
+    units: u64,
+    expiry: u64,
+}
+
+#[derive(Debug)]
+pub enum EventType {
+    Signer(SignerEvent),
+    SignerMigrated { migrated_at: u64 },
+    IdRegister(IdRegisterEvent),
+    StorageRent(StorageRentEvent),
+}
+
+#[derive(Debug)]
+pub struct Event {
+    chain_id: u64,
+    block_number: u64,
+    block_hash: FixedBytes<32>,
+    block_timestamp: u64,
+    log_index: u64,
+    fid: u64,
+    tx_index: u64,
+    version: u64,
+    event_type: EventType,
+}
+
+pub struct Subscriber {
+    provider: RootProvider<Http<Client>>,
+    onchain_events_by_block: HashMap<u64, Vec<Event>>,
+}
+
+// TODO(aditi): Wait for 1 confirmation before "committing" an onchain event.
+impl Subscriber {
+    pub fn new(config: Config) -> Result<Subscriber, SubscribeError> {
+        let url = config.rpc_url.parse()?;
+        let provider = ProviderBuilder::new().on_http(url);
+        return Ok(Subscriber {
+            provider,
+            onchain_events_by_block: HashMap::new(),
+        });
+    }
+
+    pub fn add_onchain_event(
+        &mut self,
+        fid: u64,
+        block_number: u64,
+        block_hash: FixedBytes<32>,
+        block_timestamp: u64,
+        log_index: u64,
+        tx_index: u64,
+        event_type: EventType,
+    ) {
+        let event = Event {
+            fid,
+            block_number,
+            block_hash,
+            block_timestamp,
+            log_index,
+            tx_index,
+            event_type,
+            chain_id: CHAIN_ID,
+            version: 0,
+        };
+        let events = self.onchain_events_by_block.get_mut(&block_number);
+        match events {
+            None => {
+                self.onchain_events_by_block
+                    .insert(block_number, vec![event]);
```

Comment:
Is this setting the key to the empty vec if it doesn't exist? Do we also have to push the event in this case? If not, perhaps a comment to explain?

Reply:
I might be misunderstanding this comment, but we do insert the event. The vec is initialized with the event inside of it.

Reply:
Yes, missed that thanks. Still getting used to reading rust.